common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala

```diff
@@ -15,44 +15,81 @@ class SQSReaderTest
     val sqsConfig =
-      SQSConfig("eu-west-1",
-                idMinterQueueUrl,
-                waitTime = 20 seconds,
-                maxMessages = 2)
+      SQSConfig("eu-west-1", queueUrl, waitTime = 20 seconds, maxMessages = 2)
     val messageStrings = List("someMessage1", "someMessage2", "someMessage3")
-    messageStrings.foreach(sqsClient.sendMessage(idMinterQueueUrl, _))
+    messageStrings.foreach(sqsClient.sendMessage(queueUrl, _))
     val sqsReader =
       new SQSReader(sqsClient, sqsConfig)
 
-    val futureMessages = sqsReader.retrieveMessages()
+    var receivedMessages: List[Message] = Nil
 
-    whenReady(futureMessages) { messages =>
+    val futureMessages = sqsReader.retrieveAndDeleteMessages(message => receivedMessages = message :: receivedMessages)
+
+    whenReady(futureMessages) { _ =>
       // SQS is not a FIFO queue and it only guarantees that a message is sent at least once,
       // not that it is received exactly once
-      messages should have size 2
-      messages.foreach { message =>
+      receivedMessages should have size 2
+      receivedMessages.foreach { message =>
         messageStrings should contain(message.getBody)
       }
     }
+
+    // Check that the previous 2 messages have been deleted
+    assertNumberOfMessagesAfterVisibilityTimeoutIs(1, sqsReader)
   }
 
-  it("should return a failed future if writing to the SNS topic fails") {
+  it("should return a failed future if reading from the SQS queue fails") {
     val sqsConfig =
       SQSConfig("eu-west-1",
                 "not a valid queue url",
                 waitTime = 20 seconds,
                 maxMessages = 1)
     val sqsReader = new SQSReader(sqsClient, sqsConfig)
 
-    val futureMessages = sqsReader.retrieveMessages()
+    val futureMessages = sqsReader.retrieveAndDeleteMessages(_ => ())
 
     whenReady(futureMessages.failed) { exception =>
       exception.getMessage should not be (empty)
     }
   }
 
-  private def createMessage(jsonMessage: String) =
-    new Message().withBody(jsonMessage)
+  it("should return a failed future if processing one of the messages fails - the failed message should not be deleted") {
+    val sqsConfig =
+      SQSConfig("eu-west-1", queueUrl, waitTime = 20 seconds, maxMessages = 10)
+
+    val failingMessage = "failingMessage"
+    val messageStrings = List("someMessage1", failingMessage, "someMessage3")
+    messageStrings.foreach(sqsClient.sendMessage(queueUrl, _))
+    val sqsReader =
+      new SQSReader(sqsClient, sqsConfig)
+
+    val futureMessages = sqsReader.retrieveAndDeleteMessages { message =>
+      if (message.getBody == failingMessage)
+        throw new RuntimeException(s"$failingMessage is not valid")
+      else message
+    }
+
+    whenReady(futureMessages.failed) { exception =>
+      exception shouldBe a[RuntimeException]
+    }
+
+    // Check that the queue still contains all 3 messages
+    assertNumberOfMessagesAfterVisibilityTimeoutIs(1, sqsReader)
```

Comment:
Oops, that is one of the reasons I don't like comments. I'll write a self explanatory name for that assertion and delete the comment